# Phase 1: Data Foundation — Step 1.1: Data Understanding

This notebook performs the initial data understanding for the 5 active raw datasets of the Agentic HRMS project, reporting their shape, data types, missing value counts, duplicate counts, any ID-related columns, and the class balance for employee attrition.

In [1]:
import os
import pandas as pd

raw_dir = os.path.join("data", "raw")
print(f"Checking files in: {os.path.abspath(raw_dir)}")

Checking files in: C:\Users\Harshit Mishra\OneDrive\Desktop\enterprise_hr_ai\data\raw


## 1. Employee Attrition Dataset (`employee_attrition.csv`)
* **Role**: Attrition ML source + Employee Master
* **Description**: Feeds the attrition model and stands in as the core employee table.

In [2]:
file_path = os.path.join(raw_dir, "employee_attrition.csv")
df_attrition = pd.read_csv(file_path)

print(f"Shape: {df_attrition.shape}")
print("\nData Types:")
print(df_attrition.dtypes)

print("\nMissing Values:")
missing = df_attrition.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values")

print(f"\nDuplicate Count: {df_attrition.duplicated().sum()}")

id_cols = [c for c in df_attrition.columns if "id" in c.lower()]
print(f"Columns containing 'id': {id_cols}")

print("\nAttrition Class Balance:")
print(df_attrition["Attrition"].value_counts())
print(df_attrition["Attrition"].value_counts(normalize=True))

Shape: (1470, 35)

Data Types:
Age                         int64
Attrition                     str
BusinessTravel                str
DailyRate                   int64
Department                    str
DistanceFromHome            int64
Education                   int64
EducationField                str
EmployeeCount               int64
EmployeeNumber              int64
EnvironmentSatisfaction     int64
Gender                        str
HourlyRate                  int64
JobInvolvement              int64
JobLevel                    int64
JobRole                       str
JobSatisfaction             int64
MaritalStatus                 str
MonthlyIncome               int64
MonthlyRate                 int64
NumCompaniesWorked          int64
Over18                        str
OverTime                      str
PercentSalaryHike           int64
PerformanceRating           int64
RelationshipSatisfaction    int64
StandardHours               int64
StockOptionLevel            int64
TotalWorkingYears

## 2. HR Performance & Engagement Dataset (`hr_performance_engagement.csv`)
* **Role**: Performance & Engagement source
* **Description**: Has engagement, satisfaction, training details, and work-life balance scores.

In [3]:
file_path = os.path.join(raw_dir, "hr_performance_engagement.csv")
df_perf = pd.read_csv(file_path)

print(f"Shape: {df_perf.shape}")
print("\nData Types:")
print(df_perf.dtypes)

print("\nMissing Values:")
missing = df_perf.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values")

print(f"\nDuplicate Count: {df_perf.duplicated().sum()}")

id_cols = [c for c in df_perf.columns if "id" in c.lower()]
print(f"Columns containing 'id': {id_cols}")

Shape: (2845, 28)

Data Types:
Employee ID                     int64
StartDate                         str
Title                             str
BusinessUnit                      str
EmployeeStatus                    str
EmployeeType                      str
PayZone                           str
EmployeeClassificationType        str
DepartmentType                    str
Division                          str
DOB                               str
State                             str
GenderCode                        str
RaceDesc                          str
MaritalDesc                       str
Performance Score                 str
Current Employee Rating         int64
Survey Date                       str
Engagement Score                int64
Satisfaction Score              int64
Work-Life Balance Score         int64
Training Date                     str
Training Program Name             str
Training Type                     str
Training Outcome                  str
Training Duration(D

## 3. Occupation Data Dataset (`occupation_data.csv`)
* **Role**: Role & skills taxonomy (O*NET-SOC codes to titles)
* **Description**: Maps O*NET-SOC codes to job titles and descriptions.

In [4]:
file_path = os.path.join(raw_dir, "occupation_data.csv")
df_occ = pd.read_csv(file_path)

print(f"Shape: {df_occ.shape}")
print("\nData Types:")
print(df_occ.dtypes)

print("\nMissing Values:")
missing = df_occ.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values")

print(f"\nDuplicate Count: {df_occ.duplicated().sum()}")

id_cols = [c for c in df_occ.columns if "id" in c.lower()]
print(f"Columns containing 'id': {id_cols}")

Shape: (1016, 3)

Data Types:
O*NET-SOC Code    str
Title             str
Description       str
dtype: object

Missing Values:
No missing values

Duplicate Count: 0
Columns containing 'id': []


## 4. Essential Skills Dataset (`essential_skills.csv`)
* **Role**: Role & skills taxonomy (essential skills details)
* **Description**: Contains O*NET-SOC codes mapped to element/skill IDs, names, scales, and data values.

In [5]:
file_path = os.path.join(raw_dir, "essential_skills.csv")
df_ess = pd.read_csv(file_path)

print(f"Shape: {df_ess.shape}")
print("\nData Types:")
print(df_ess.dtypes)

print("\nMissing Values (Top 20):")
missing = df_ess.isnull().sum()
print(missing[missing > 0].head(20) if missing.sum() > 0 else "No missing values")

print(f"\nDuplicate Count: {df_ess.duplicated().sum()}")

id_cols = [c for c in df_ess.columns if "id" in c.lower()]
print(f"Columns containing 'id': {id_cols}")

Shape: (18200, 15)

Data Types:
O*NET-SOC Code            str
Title                     str
Element ID                str
Element Name              str
Scale ID                  str
Scale Name                str
Data Value            float64
N                       int64
Standard Error        float64
Lower CI Bound        float64
Upper CI Bound        float64
Recommend Suppress        str
Not Relevant              str
Date                      str
Domain Source             str
dtype: object

Missing Values (Top 20):
Not Relevant    9100
dtype: int64

Duplicate Count: 0
Columns containing 'id': ['Element ID', 'Scale ID']


## 5. Software Skills Dataset (`software_skills.csv`)
* **Role**: Role & skills taxonomy (software tools/skills details)
* **Description**: Maps O*NET-SOC codes to specific software skills/workplace examples.

In [6]:
file_path = os.path.join(raw_dir, "software_skills.csv")
df_soft = pd.read_csv(file_path)

print(f"Shape: {df_soft.shape}")
print("\nData Types:")
print(df_soft.dtypes)

print("\nMissing Values:")
missing = df_soft.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else "No missing values")

print(f"\nDuplicate Count: {df_soft.duplicated().sum()}")

id_cols = [c for c in df_soft.columns if "id" in c.lower()]
print(f"Columns containing 'id': {id_cols}")

Shape: (31821, 7)

Data Types:
O*NET-SOC Code       str
Title                str
Workplace Example    str
Element ID           str
Element Name         str
Hot Technology       str
In Demand            str
dtype: object

Missing Values:
No missing values

Duplicate Count: 0
Columns containing 'id': ['Element ID']


## 6. Summary of Data Reality

| Dataset File | Row Count | Col Count | Likely Primary Key | % Missing Cells | Purpose (from Roadmap) |
|---|---|---|---|---|---|
| `employee_attrition.csv` | 1,470 | 35 | `EmployeeNumber` | 0.00% | Attrition ML source + Employee Master (feeds attrition model, acts as core employee table). |
| `hr_performance_engagement.csv` | 2,845 | 28 | `Employee ID` | 0.00% | Performance / Engagement analytics source (keeps survey/engagement and training data). |
| `occupation_data.csv` | 1,016 | 3 | `O*NET-SOC Code` | 0.00% | Role & skills taxonomy mapping codes to job titles and descriptions. |
| `essential_skills.csv` | 18,200 | 15 | Composite (O*NET-SOC Code, Element ID, Scale ID) | 3.33% | Role & skills taxonomy mapping O*NET codes to essential skill values. |
| `software_skills.csv` | 31,821 | 7 | Composite (O*NET-SOC Code, Element ID, Workplace Example) | 0.00% | Role & skills taxonomy mapping O*NET codes to software tools. |